source/FormationPattern.cpp

```diff
@@ -0,0 +1,508 @@
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "FormationPattern.h"
+#include "Ship.h"
+
+
+#include <cmath>
+
+using namespace std;
+
+
+
+void FormationPattern::ActiveFormation::Tally(const Ship &ship)
+{
+	maxDiameter = max(maxDiameter, ship.Radius() * 2.);
+	maxHeight = max(maxHeight, ship.Height());
+	maxWidth = max(maxWidth, ship.Width());
+}
+
+
+
+void FormationPattern::ActiveFormation::ClearParticipants()
+{
+	maxDiameter = 0.;
+	maxHeight = 0.;
+	maxWidth = 0.;
+}
+
+
+
+FormationPattern::PositionIterator::PositionIterator(const FormationPattern &pattern,
+	const FormationPattern::ActiveFormation &af, unsigned int startRing, unsigned int shipsToPlace)
+	: activeFormation(af), pattern(pattern), ring(startRing), shipsToPlace(shipsToPlace)
+{
+	MoveToValidPosition();
+}
+
+
+
+const Point *FormationPattern::PositionIterator::operator->()
+{
+	return &currentPoint;
+}
+
+
+
+const Point &FormationPattern::PositionIterator::operator*()
+{
+	return currentPoint;
+}
+
+
+
+FormationPattern::PositionIterator &FormationPattern::PositionIterator::operator++()
+{
+	if(!atEnd)
+		slot++;
+	// Number of ships is used as number of remaining ships still to be placed.
+	if(shipsToPlace > 0)
+		--shipsToPlace;
+	MoveToValidPosition();
+	return *this;
+}
+
+
+
+unsigned int FormationPattern::PositionIterator::Ring() const
+{
+	return ring;
+}
+
+
+
+void FormationPattern::PositionIterator::MoveToValidPosition()
+{
+	// If we cannot calculate any new positions, then just return center point.
+	if(atEnd)
+	{
+		currentPoint = Point();
+		return;
+	}
+	
+	// Check if there are any lines available.
+	unsigned int lines = pattern.get().Lines();
+	if(lines < 1)
+		atEnd = true;
+	
+	unsigned int ringsScanned = 0;
+	unsigned int startingRing = ring;
+	unsigned int lineRepeatSlots = pattern.get().Slots(ring, line, repeat);
+	
+	while(slot >= lineRepeatSlots && !atEnd)
+	{
+		unsigned int patternRepeats = pattern.get().Repeats(line);
+		// LineSlot number is beyond the amount of slots available.
+		// Need to move a ring, a line or a repeat-section forward.
+		if(ring > 0 && line < lines && patternRepeats > 0 && repeat < patternRepeats - 1)
+		{
+			// First check if we are on a valid line and have another repeat section.
+			++(repeat);
+			slot = 0;
+			lineRepeatSlots = pattern.get().Slots(ring, line, repeat);
+		}
+		else if(line < lines - 1)
+		{
+			// If we don't have another repeat section, then check for a next line.
+			++(line);
+			repeat = 0;
+			slot = 0;
+			lineRepeatSlots = pattern.get().Slots(ring, line, repeat);
+		}
+		else
+		{
+			// If we checked all lines and repeat sections, then go for the next ring.
+			++(ring);
+			line = 0;
+			repeat = 0;
+			slot = 0;
+			lineRepeatSlots = pattern.get().Slots(ring, line, repeat);
+			
+			// If we scanned more than 5 rings without finding a slot, then we have an empty pattern.
+			++ringsScanned;
```

Comment:
Might as well be consistent: Missing parens here and in some other places too.

Reply:
The original code was a more complex and needed parens, I will remove them above since they don't add value here.